File: Python/Easy/Trees/Symmetric_Tree.py
## Author: Chris Hoerle
## Date: 08/12/2021

'''
	Given the root of a binary tree, check whether it is a mirror 
	of itself (i.e., symmetric around its center).

 	Examples:
	Input: root = [1,2,2,3,4,4,3]
	Output: true

	Input: root = [1,2,2,null,3,null,3]
	Output: false
 
	Constraints:
	The number of nodes in the tree is in the range [1, 1000].
	-100 <= Node.val <= 100
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True
        
        return self.symmetric(root.left, root.right)
    
    
    def symmetric(self, left, right):
        if left is None or right is None:
            if left == right:
                return True
            return False
        
        if left.val != right.val:
            return False
        
        return self.symmetric(left.left, right.right) and self.symmetric(left.right, right.left)

File: Java/Easy/Trees/Symmetric_Tree.java
// Author: Chris Hoerle
// Date: 08/12/2021

/*
	Given the root of a binary tree, check whether it is a mirror 
	of itself (i.e., symmetric around its center).

 	Examples:
	Input: root = [1,2,2,3,4,4,3]
	Output: true

	Input: root = [1,2,2,null,3,null,3]
	Output: false
 
	Constraints:
	The number of nodes in the tree is in the range [1, 1000].
	-100 <= Node.val <= 100
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Time Complexity: O(n) because need to traverse through entire tree
class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;   
        }  
        
        return symmetric(root.left, root.right);
    }
    
    public boolean symmetric(TreeNode left, TreeNode right) {
        if (left == null || right == null) {
            if (left == right) {
                return true;
            }
            return false;
        }
        
        if (left.val != right.val) {
            return false;
        }
        
        return symmetric(left.left, right.right) && symmetric(left.right, right.left); 
    }
}